docs/design/resource-decoding.md

```diff
@@ -0,0 +1,322 @@
+
+## Non-Goals
+
+- Encoding Objects
+
+## Design Components
+
+### **Patches**
+
+```go
+type MutateFunc func(k8s.Object) error
+```
+
+All decoding functions accept a `patches ...MutateFunc` argument. Patches may be used to inject data after decoding is complete. If a MutateFunc returns an error, decoding is halted.
+
+This may be done to inject dynamic data that may not be known until runtime or that may be sensitive like a locally valid credential.
+
+Example pre-defined MutateFuncs:
+
+```go
+// apply an override set of labels to a decoded object
+func MutateLabels(overrides map[string]string) MutateFunc
+// apply an override set of annotations to a decoded object
+func MutateAnnotations(overrides map[string]string) MutateFunc
+// apply an owner annotation to a decoded object
+func MutateOwnerAnnotations(owner k8s.Object) MutateFunc
+```
+
+### **Handlers**
+
+Some decoding functions accept a HandlerFunc, a function that is executed after decoding and the optional patches are completed per each object.
+
+If a HandlerFunc returns an error, decoding is halted.
+
+```go
+type HandlerFunc func(context.Context, k8s.Object) error
+```
+
+Example pre-defined HandlerFuncs:
+
+```go
+// CreateHandler returns a HandlerFunc that will create objects
+func CreateHandler(*Resources, opts ...CreateOption) HandlerFunc
+// UpdateHandler returns a HandlerFunc that will update objects
+func UpdateHandler(*Resources, opts ...UpdateOption) HandlerFunc
+// DeleteHandler returns a HandlerFunc that will delete objects
+func DeleteHandler(*Resources, opts ...DeleteOption) HandlerFunc
+
+// IgnoreErrorHandler returns a HandlerFunc that will ignore an error
+func IgnoreErrorHandler(HandlerFunc, error) HandlerFunc
+
+// CreateIfNotExistsHandler returns a HandlerFunc that will create objects if they do not already exist
+func CreateIfNotExistsHandler(*Resources, opts ...CreateOption) HandlerFunc
+```
+
+### Decoding a single-document YAML/JSON input
+
+The following are proposed function signatures for decoding input that contain a single `k8s.Object` type. Example:
+
+```yaml
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: my-serivce-account
+  namespace: myappns
+```
+
+1. Decoding an object to a known type
+
+```go
+func Decode(manifest io.Reader, obj k8s.Object, patches ...MutateFunc) error
+```
+
+Usage:
+
+```go
+sa := v1.ServiceAccount{}
+err := Decode(strings.NewReader("..."), &sa)
+```
+
+With a MutateFunc:
+
+```go
+// Decode to sa and apply the label "test" : "feature-X"
+sa := v1.ServiceAccount{}
+err := Decode(strings.NewReader("..."), &sa, MutateLabels(map[string]string{"test" : "feature-X"}))
+```
+
+2. Decoding an object without knowing the type
+
+`defaults` is an optional parameter, if specified, it is a hint to the decoder to help determine the underlying Go type to use for object creation.
+
+```go
+func DecodeAny(manifest io.Reader, defaults *schema.GroupVersionKind, patches ...MutateFunc) (k8s.Object, error)
```

Comment:
Could the GVK be passed in as a MutateFunc? That way the method signature is even simpler.

Reply:
MutateFunc wouldn't be my first choice, as it has this signature: `type MutateFunc func(obj k8s.Object) error` -- but I could try to collapse the two into a DecodeOption  func/type that can be used in this variadic parameter.

That sounds like a good idea